[9.0][ADD] attendance sheets

elouan-lb opened the pull request:
Extends `beesdoo_shift` module with an `attendance sheets` based functionnality. It covers two different uses : one on a generic computer (group `attendance_sheet_access`) and the other one for administrators (group `cooperative_admin`)

## Sheets generation

The attendance sheets are automatically generated from CRON `ir_cron_generate_attendance_sheet`. The system checks if any shift starts within the time intervall defined in settings and creates the corresponding sheet, pre-filled with already expected shifts.

 ## Shifts

The sheet stores information from the time it has been validated, that's why its shifts are **copies** from actual shifts in the system, which are updated only at validation. A validated sheet cannot be edited anymore.

## Card scan 

The system supports members cards scan. When a cooperator scans its card, his shift is updated to _done_ if he was expected. If he wasn't, a new `added_shift` is created.

## Validation 

Only supercooperators and `cooperative_admin` are allowed to validate an attendance sheet. They must scan their card or login if the sheet is displayed with a generic account (wizard `beesdoo.shift.sheet.validate`). 

Emails are sent to absent users on validation.

Added shifts are created only if no empty shift (without any worker) exists for the same `Task Type`. These are updated firstly.

## Sheets status

Attendance sheets are displayed in dedicated worklists when **non-validated** or **annotated** (the supercooperator let a note to be treated by Members Office). 

The CRON `ir_cron_check_non_validated_sheet` send an email to Members Office if any sheet from past day is still non-validated.

## Unite testing

Unite tests are added to the module, to automatically test shifts status, cooperators status and attendance sheets.

Commit:
[ADD] b_shift : attendance sheet models

Commit:
[ADD] b_shift : attendance_sheet_id to TaskTemplate

Commit:
[ADD] b_shift : attendance sheet views

Commit:
[ADD] b_shift : wizard for attendance sheet generation

Commit:
[ADD] b_shift : attendance sheet access rules

Commit:
[FIX] b_easy_my_coop: change dependencies order

Commit:
[ADD] b_shift : attendance sheet validation wizard

Commit:
[ADD] b_shift : attendance sheet notifications count

Commit:
[ADD] b_shift : attendance sheet list items coloration

Commit:
[FIX] b_shift : attendance sheet replacement worker field name

Commit:
[IMP] b_shift : attendance sheet views set no_open options

Commit:
[IMP] b_shift : attendance sheet display date with ISO format

Commit:
[ADD] b_shift : attendance shift validation barcode field

Commit:
[ADD] b_shift : default task type configuration

Commit:
[IMP] b_shift : constraint on validated attendance sheets

Commit:
[ADD] b_shift : attendance sheet "mark as read" smart button

Commit:
[ADD] b_shift : attendance sheet archived attribute

Commit:
[ADD] b_shift : demo datas

Commit:
[ADD] b_base : demo datas

Commit:
[ADD] b_base : demo datas worker eater

Commit:
[ADD] b_shift : attendance sheets barcode scanner support

Commit:
[REM] b_shift : attendance sheet cancelled status

Commit:
[FIX] b_shift : attendance sheet shifts worker not required

Commit:
[IMP] b_shift : attendance sheets linting and menu sequences

Commit:
[FIX] b_shift : comment attendance sheet unique worker constraint

Commit:
[IMP] b_shift : change and add access rules

Commit:
[IMP] b_shift : attendance sheets code cleaning

Commit:
[FIX] b_shift : constrain_unique_worker on attendance sheets

Commit:
[FIX] b_shift : code cleaning

Commit:
[IMP] b_shift : attendance sheet validation exceptions raised

Commit:
[ADD] b_shift : barcode scan to attendance sheet validation

Commit:
[FIX] b_shift : attendance sheet compensation_nb onchange issues

The @api.onchange, modifying compensation_nb according to working mode,
led to strange behavior and lack of consistency.

compensation_nb has consequently moved in Shift Stage.

Commit:
[FIX] b_shift : barcode scan to attendance sheet validation

Commit:
[REM] beesdoo_shift_assets

Commit:
[FIX] b_shift : attendance sheet scanning methods

Commit:
[ADD] b_shift : constraint on irregular_start_date

Commit:
[IMP] b_shift : attendance sheet form view

Commit:
[REM] b_shift : expected_worker_nb and added_worker_nb from attendance sheet

They are already automatically displayed on the form view.

Commit:
[IMP] b_shift : attendance sheet validation button hidden when validated

Commit:
[ADD] b_shift : attendance sheet possibility to translate terms

Commit:
[FIX] b_shift : date format %%Y-%%m-%%d

Commit:
[ADD] b_shift : attendance sheet CRON for generation

Commit:
[FIX] b_shift : code cleaning

Commit:
[REM] b_shift : generate_attendance_sheet views

Commit:
[ADD] b_shift : attendance sheet default super cooperator

Commit:
[ADD] b_shift : attendance sheet search view

Commit:
[FIX] b_shift : computed methods not applying to recordset

Commit:
[ADD] b_shift : attendance sheet time slot

Commit:
[ADD] b_shift : demo datas

Commit:
[ADD] b_easy_my_coop : share_lines demo datas

Commit:
Revert "[ADD] b_shift : demo datas"

This reverts commit 6c0d0a0d8c2952628c4d26d0acc4d115f042e118.

Commit:
[FIX] b_shift : stored field in abstract class

Commit:
[FIX] b_shift : attendance sheet, validated by

Commit:
[FIX] b_shift : attendance sheet annotation validation

Commit:
[FIX] b_shift : wrong default regular task type

Commit:
[REF] b_shift : attendance sheet code cleaning

Commit:
[ADD] b_shift : attendance sheet automatic emails

Commit:
[IMP] b_shift : change attendance sheets shifts order

Commit:
[ADD] b_w_shift : prevent subscription 20 min before start

Commit:
[IMP] b_shift : change configuration menu order

Commit:
[ADD] b_shift : attendance sheet tests setUp

Commit:
[FIX] b_shift : search on equal dates

Commit:
[ADD] b_shift : attendance sheets test on creation

Commit:
[ADD] b_shift : attendance sheet test on default task type

Commit:
[ADD] b_shift : member cards demo datas

Commit:
[ADD] b_shift : exception raised when replaced worker try to subscribe

Commit:
[ADD] b_shift : attendance sheet unit test on barcode scanner

Commit:
[IMP] b_shift : no exception raised when scanning a worker twice

Commit:
[IMP] b_shift : cancelled shift not added to attendance sheets

Commit:
[IMP] b_shift : attendance sheet max worker no calculation

Now based on created shifts instead of task templates

Commit:
Revert "[ADD] b_shift : attendance sheet default super cooperator"

This reverts commit 605ac1853f6ab0497bb0a6bf566c66655056269d.

Commit:
[REM] b_shift : attendance sheet shift 'cancelled' status

Commit:
[REF] b_shift : code cleaning

Commit:
[IMP] b_shift : cooperative admin can validate sheet with their cards

Commit:
[FIX] b_shift : cron for non-validated sheets

Commit:
[UPD] b_shift : attendance sheet name splitted in day and time_slot on view

Commit:
[IMP] b_shift : attendance annotation pre-filled on wizard

Commit:
[REM] b_shift : Shift Attendance view

Commit:
[IMP] b_shift : group_shift_attendance can accessShift menu

Commit:
[IMP] b_shift : attendance sheet generation time configurable

Generation CRON is running every minute and check for sheets
in the configurable time interval. Workers can't subscribe
to a shift on intranet if its start time is in this interval.

Commit:
[ADD] b_shift : attendance sheet, is_compensation field for expected shifts

Commit:
[ADD] b_shift : week name on attendance sheet

Commit:
[WIP] b_shift : give access to backend to attendance sheets generic group

Commit:
[ADD] b_shift : attendance sheet compensation number

Changes from shift status update are made as well.

Commit:
[FIX] b_shift : time on added shift

Commit:
[REF] b_shift : code cleaning

Commit:
[IMP] b_shift : automatically created shifts names

Commit:
[CHG] b_shift : checks on attendance sheet before validation

Commit:
[REF] b_shift : views order and definition

Commit:
[ADD] b_shift : replacement worker on shift kanban view

Commit:
[ADD] b_shift : optional validation by card scan

Commit:
[ADD] b_shift : constrains on future attendance sheets validation

Commit:
[ADD] b_shift : day_abbreviation for attendance sheets

Commit:
[FIX] b_shift : cron no_update='1'

Commit:
[FIX] b_shift : attendance sheet 'is_annotated'

Commit:
[IMP] b_shift : attendance sheet validation wizard

Temporary changes can be saved.
Warnings added and modified.

Commit:
[FIX] b_shift : sheet generation cron overloading ram

Commit:
[REF] b_shift : code cleaning

Commit:
[ADD] b_shift : _order for attendance sheet shifts

Commit:
[REF] b_shift : code cleaning

Commit:
[ADD] b_shift : group_by on worker_no_feedback

Commit:
[ADD] b_shift : demo datas for task templates

Commit:
[FIX] b_shift : supercoop name not displayed on sheets

Commit:
[UPD] b_shift : demos datas & unite tests

Commit:
[CHG] b_shift : attendance sheets terminology

elouan-lb commented on beesdoo_shift/models/attendance_sheet.py:
```diff
@@ -0,0 +1,634 @@
+        }
+
+    @api.model
+    def _generate_attendance_sheet(self):
```
Complexity of this function is not ideal, would appreciate some feedback !

robinkeunen replied:
Je vois pas le problème. :/

elouan-lb replied:
C'est une méthode appelée par un CRON toutes les 3 minutes, avec plusieurs appels à `search()`. Comme l'intervalle de génération (15 minutes par défaut) est plus grand que la fréquence d'exécution du CRON, ça fait beaucoup d'appels pour rien.

remytms commented on beesdoo_shift/data/mail_template.xml:
```diff
@@ -1,8 +1,95 @@
+    <br/><br/> You were supposed to replace ${object.worker_id.name}. <br/><br/>
+    % endif
+
+    % if object.replaced_id:
```
I think it should be `% if object. worker_id:`

elouan-lb replied:
Indeed, thank you !

remytms commented on beesdoo_shift/models/attendance_sheet.py:
```diff
@@ -0,0 +1,634 @@
+        id = (
+            int(parameters.get_param("beesdoo_shift.default_task_type_id"))
+            or 1
+        )
```
Here you can use : 
```python
id = int(parameters.get_param("beesdoo_shift.default_task_type_id", default=1))
```

remytms commented on beesdoo_shift/models/attendance_sheet.py:
```diff
@@ -0,0 +1,634 @@
+        replacement_ids = map(
+            lambda s: s.replacement_worker_id.id, self.expected_shift_ids
+        )
+        replacement_ids = filter(bool, replacement_ids)
```
Nothing to change, but just to troll: I find list comprehension more pythonic.
```python
added_ids = (s.worker_id.id for s in self.added_shift_ids)
expected_ids = (s.worker_id.id for s in self.expected_shift_ids)
replacement_ids = (s.replacement_worker_id.id for s in self.expected_shift_ids)
replacement_ids = (bool(s) for s in replacement_ids)
```
You can use braces instead of parenthesis if you prefer list instead of generators.

@robinkeunen will certainly disagree with me. ;)

robinkeunen replied:
Nope, I agree, list comprehensions are indeed more pythonic.

I disagree on the filter though: 

```python
replacement_ids = [
    s.replacement_worker_id.id 
    for s in self.expected_shift_ids 
    if s.replacement_worker_id.id
]
```

robinkeunen replied:
Nope, I agree, list comprehensions are more pythonic. I disagree on the filter though:

```python
replacement_ids = [
    s.replacement_worker_id.id 
    for s in self.expected_shift_ids
    if s.replacement_worker_id.id
]
```

elouan-lb replied:
That's right, thank you :)

remytms commented on beesdoo_shift/models/attendance_sheet.py:
```diff
@@ -0,0 +1,634 @@
+                "worker_id": worker.id,
+                "is_compensation": is_compensation,
+            }
+        )
```
Not sure if I understand it well, but I would prefer:

```python
if worker.id not in added_ids:
        # added shift creation
        self.added_shift_ids |= self.added_shift_ids.new(
            {
                "task_type_id": self.added_shift_ids.default_task_type_id(),
                "state": "done",
                "attendance_sheet_id": self._origin.id,
                "worker_id": worker.id,
                "is_compensation": is_compensation,
            }
        )
```

remytms commented on beesdoo_shift/models/cooperative_status.py:
```diff
@@ -201,6 +201,11 @@ def _compute_next_countdown_date(self):
                         next_countdown_date = date
                 rec.next_countdown_date = next_countdown_date
 
+    @api.constrains("working_mode", "irregular_start_date")
+    def _constrains_irregular_start_date(self):
+        if self.working_mode == "irregular" and not self.irregular_start_date:
+            raise UserError(_("Irregular workers must have an irregular start date."))
+
```
Are you sure about that? I have not looked more, but is'nt the `start_date` removed when unsubscribed ?

elouan-lb replied:
It's not removed !

remytms commented on beesdoo_shift/models/res_config.py:
```diff
@@ -0,0 +1,74 @@
+                self.env["ir.config_parameter"].get_param(
+                    "beesdoo_shift.card_support"
+                )
+                == "True"
```
For that I use: https://docs.python.org/3/library/ast.html?highlight=literal%20eval#ast.literal_eval

elouan-lb replied:
Nice, thanks :)

remytms commented on beesdoo_shift/security/group.xml:
```diff
@@ -17,7 +21,18 @@
+
+        <!-- <template id="website.user_navbar">
+            <field name="groups"  eval="[(6,0,[ref('group_shift_attendance_sheet')])]"/>
+        </template> -->
```
Remove it, if it's not used. :)

remytms commented on beesdoo_shift/wizard/validate_attendance_sheet.py:
```diff
@@ -0,0 +1,138 @@
+            self.env["ir.config_parameter"].get_param(
+                "beesdoo_shift.card_support"
+            )
+            == "True"
```
Same as above.

remytms commented on beesdoo_shift/wizard/validate_attendance_sheet.py:
```diff
@@ -0,0 +1,138 @@
+            if not self.login:
+                raise UserError(_("Please enter your login."))
+            user = self.env["res.users"].search([("login", "=", self.login)])
+            user.sudo(user.id).check_credentials(self.password)
```
You are not checking result of `check_credentials`. Maybe it only raise errors ? I have'nt read ``check_credentials` definition. It's just to be sure.

elouan-lb replied:
Indeed, I only use it to raise an exception if credentials are wrong or if user has wrong rights.

remytms reviewed (changes requested):
Fameuse relecture. :) 

Très bon travail ! Une petite correction, sinon quelques remarques mais qui ne nécessite pas d'office un changement.

robinkeunen commented on beesdoo_shift/models/res_config.py:
```diff
@@ -0,0 +1,74 @@
+                self.env["ir.config_parameter"].get_param(
+                    "beesdoo_shift.card_support"
+                )
+                == "True"
```
C'est déjà un booléen, pas besoin de tester avec "== True".

remytms replied:
C'est des string en fait, je pense. Donc pour moi c'est un `ast.literal_eval` qu'il faut faire. Mais à vérifier.

robinkeunen replied:
yes, mon mauvais.

robinkeunen commented on beesdoo_shift/wizard/validate_attendance_sheet.py:
```diff
@@ -0,0 +1,138 @@
+            self.env["ir.config_parameter"].get_param(
+                "beesdoo_shift.card_support"
+            )
+            == "True"
```
Même commentaire que ci-dessus. Si tu as peur de recevoir un `None`,  tu peux écrire `bool(<expression>)`.

robinkeunen replied:
(j'ai rien dit, le paramètre est une string. Pourquoi?)

elouan-lb replied:
C'est la façon dont les paramètres sont stockés, j'en ai bien l'impression. Peut-être que @remytms saura t'en dire plus !

remytms replied:
> (j'ai rien dit, le paramètre est une string. Pourquoi?)

Pour être générique, on ne peut stocker que des string dans la table `ir_config_parameter`. C'est une table clé-valeur toute simple partagée par tous les modules. :)

robinkeunen commented on beesdoo_shift/wizard/validate_attendance_sheet.py:
```diff
@@ -0,0 +1,138 @@
+    Useless for users in group_cooperative_admin"""
+    _inherit = ["barcodes.barcode_events_mixin"]
+
+    @api.multi
+    def _get_active_sheet(self):
+        sheet_id = self._context.get("active_id")
+        sheet_model = self._context.get("active_model")
```
Why didn't you set `active_sheet` as a field. Access would be simpler to write and would avoid the browse for each access to its field (even if most probably cached).

elouan-lb replied:
It's true ! Changes are done, I don't know why I used this painful way.

robinkeunen commented on beesdoo_shift/wizard/validate_attendance_sheet.py:
```diff
@@ -0,0 +1,138 @@
+        """
+        sheet = self._get_active_sheet()
+
+        sheet.notes = self.notes
+        sheet.feedback = self.feedback
+        sheet.worker_nb_feedback = self.worker_nb_feedback
```
if active_sheet was a field, you could persist it directly.

elouan-lb replied:
Still, as it is a wizard I should not override `write` method, don't you think ?

robinkeunen commented on beesdoo_website_shift/controllers/main.py:
```diff
@@ -143,12 +145,21 @@ def subscribe_to_shift(self, shift_id=-1, **kw):
+                "beesdoo_shift.attendance_sheet_generation_interval"
+            )
+        )
+        start_time_limit = datetime.now() + timedelta(minutes = setting)
```
pas d'espaces autour des _named parameters_.

robinkeunen reviewed (changes requested):
Excellent boulot mec, parmi le meilleur code que j'ai lu à Coop IT Easy, sans exagération.

Commit:
[FIX] b_shift : search on datetime with '='

Commit:
[REF] b_shift : code cleaning

Commit:
[FIX] b_shift : attendance sheet _compute_is_annotated

Commit:
[REF] b_shift : code cleaning

robinkeunen commented:
@elouan-lb Si tu tardes trop à _squasher_ tes _fixup_, ça va virer au cauchemar ;-)

PS: RIP la langue française.

Commit:
[IMP] b_shift : attendance sheet validation wizard

elouan-lb reviewed (commented):
> @elouan-lb Si tu tardes trop à _squasher_ tes _fixup_, ça va virer au cauchemar ;-)

Rien de tel qu'un bon paquet de conflits pour se mettre en jambe, merci du conseil 

> Excellent boulot mec, parmi le meilleur code que j'ai lu à Coop IT Easy, sans exagération.

Merci beaucoup ! :)

J'ai modifié le code en accord avec vos remarques

Commit:
[ADD] b_shift : tests for cooperative status

Commit:
[ADD] b_base : demo datas for generic and permanent users

Commit:
[ADD] b_shift : authors names

Commit:
[MOV] b_shift : demo datas for users in b_base

Commit:
[CHG] b_shift : _cron_non_validated_sheets inactive by default

Commit:
[ADD] b_shift : traductions

Commit:
[FIX] b_shift : email for non-attended shifts

Commit:
[ADD] b_shift : traductions

Commit:
[FIX] b_shift : created shifts not updating counters

If one shift is directly created as "Attended",
workers' shifts counters are not updated.
They are only updated on state transitions.

Commit:
[UPD] b_shift : unite tests

Commit:
[FIX] b_shift : email_template_non_attendance

Commit:
[REF] b_shift : code cleaning

Commit:
[FIX] b_shift : error messages

Commit:
[UPD] b_shift : traductions

Commit:
[FIX] b_shift : security groups for coop admin

Commit:
[ADD] b_shift : constrains on exempted workers

Commit:
[REF] b_shift : remove state computation for attendance shifts

replacement_worker_id -> replaced_id

Commit:
[REF] b_shift : demos

Commit:
[ADD] b_shift : domain on attendance shifts

Commit:
[FIX] b_shift : automatic error message

For regular non-attended not doing
their compensation shift.

Commit:
[UPD] b_shift : security groups

Add a new groupe only for attendance
sheets validation.

Commit:
[ADD] b_shift : option for worker_nb feedback

Commit:
[UPD] b_shift : attendance sheet form view

Commit:
[FIX] b_shift : attendance sheet generation

When several same shifts exists

Commit:
fixup! [FIX] b_shift : email_template_non_attendance

Commit:
[UPD] b_shift : translations

Commit:
[UPD] b_shift : attendance sheets view

To ease scanner use.

houssine78 commented:
@elouan-lb I leave the review to @remytms as it has a better view on this module than I do !

You seems as some conflicts to resolve though

Commit:
[CHG] b_shift : generated shifts name computation

Commit:
[FIX] b_shift : default_task_type_id computation

Commit:
[ADD] b_shift : attended_worker_no

Commit:
[REF] b_shift : attendance sheet validation wizard

Commit:
[UPD] b_shift : attendance sheets views

Commit:
[UPD] b_shifts : traductions and terms

robinkeunen commented:
@elouan-lb can we delete the merged branch?
